adds DomainMatrix to Matrix.det

hyadav2k opened the pull request:
This PR implements `DomainMatrix` to slow Matrix Method `det`

Reference to Task 6 in #20987

#### Brief description of what is fixed or changed
This PR implements `DomainMatrix` for `det` which speeds up the method by a large factor of time.


<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/matrices/determinant.py:
```diff
@@ -538,6 +540,14 @@ def _per(M):
+    K = DOM.domain
+    val = DOM.det()
+    if DOM.domain != EX:
+        val = K.to_sympy(val)
```
Why is this checking for the EX domain? Is that necessary? Surely we always want to convert to Expr.

hyadav2k replied:
Sure it is unnecessary.
Thanks.

oscarbenjamin commented:
For now I think it would be better to add using DomainMatrix in det as an option along with the others so you can do something like `M.det(method='domain')` or perhaps `M.det(method='domain-ge')` since the current implementation uses Gaussian elimination but that could be changed in future.

Before enabling the use of `DomainMatrix` by default there needs to be some analysis of performance for different types of domains and sizes of matrix. We might want to use `DomainMatrix` in some cases but not others for example. Probably there isn't much advantage in using `DomainMatrix` with the `EX` domain for `det`.

hyadav2k commented:
@oscarbenjamin for now I have added it as an option with the method hint suggested by you, and also added some test cases.
For the analysis part, is there some existing tool or we need to do it manually.
Thanks.

oscarbenjamin commented:
> For the analysis part, is there some existing tool or we need to do it manually.

To begin with I would suggest timing some examples manually. The key question is can we identify particular domains for which DomainMatrix is always better? (Note that "better" is not always the same as "faster" if the outputs of different functions are not the same)

There is a benchmark suite that can be used to compare timings across different changes:
https://github.com/sympy/sympy_benchmarks

hyadav2k commented:
@oscarbenjamin, Is this good to go?
We have added it as an option and not default as of now.
Thanks.

oscarbenjamin commented:
Can you squash the commits into one?

oscarbenjamin commented:
This looks good.

To really make use of this there needs to be some analysis of timing but at least exposing the option to end users is good.

oscarbenjamin commented:
The squashing doesn't seem to have worked because this is still showing 4 commits. This is my suggestion for how to squash the commits:
```
# First SAVE YOUR WORK somewhere OUTSIDE OF GIT

git checkout 20987_domainmatrix_slow_det

# First compute the diff for this PR and save it:
git fetch upstream
git merge upstream/master
git diff upstream/master > 21052.diff

# Now restore your branch to main master branch:
git reset --hard HEAD~100    # careful, deletes your commits, back up first!
git merge upstream/master

# Now apply the diff on top and push
git apply 21052.diff
git commit -am 'feat(matrices): add domain-ge method to Matrix.det'
git push --force
```

hyadav2k committed:
feat(matrices): add domain-ge method to Matrix.det

hyadav2k commented:
@oscarbenjamin, thank you so much, I was trying from yesterday, reading the StackOverflow, but could not succeed. I will surely work on the analysis part from now on.

There is one more thing, I want to participate in `GSoC` for this year and contribute to `sympy`, but I feel intimidated about How to choose an idea for myself? My particular interests are in topics related to `solvers`, `integration`, `ode`, `series`, and `DomainMatrix`.

I am not able to start how should I choose a topic and start working on the proposal. If you could please guide or refer to some article which can simplify things for me, it will be very generous of you.

Thanks for helping as always.

oscarbenjamin commented:
I don't know how to advise. You should choose something that you understand and that you know how to implement and where you can see significant improvements for sympy. Anyway that kind of discussion should go on the mailing list rather than here.

The tests failed for an unrelated reason so I've restarted them and enabled auto-merge

hyadav2k commented:
> I don't know how to advise. You should choose something that you understand and that you know how to implement and where you can see significant improvements for sympy. Anyway that kind of discussion should go on the mailing list rather than here.
> 
Thanks for replying, I will surely post on the mailing list.

> The tests failed for an unrelated reason so I've restarted them and enabled auto-merge
>
Sure, thanks.

oscarbenjamin commented:
I just realised that this should have had a release note. I've added one here:
https://github.com/sympy/sympy/wiki/Release-Notes-for-1.8

hyadav2k commented:
@oscarbenjamin, thank you. 
I will keep that in mind for other PRs.
